### ISSUE TITLE
LossLens Demo Scripts

### ISSUE THREAD

**Issue by tiankaixie** [Vis4SciML/loss-lens#3]
# Introduction

In this video, we demonstrate our visual analytics framework LossLens for diagnosing deep learning models from a loss landscape perspective. The system consists of 3 components. First, Global Analysis shows the global structure of loss landscapes for models of interest. Second, Local Analysis provides a comprehensive visualization of local structure of loss landscapes for selected checkpoints, including a loss contour view, persistence diagram, and merge tree. Third, semi-global analysis provides pairwise checkpoint evaluation, which includes a layer-similarity matrix, and a prediction disparity view.

# Data Alternation of Vision Transformer

In the first case study, we demonstrate how our framework supports diagnosing the influence of using the same amount of CIFAR10 and CIFAR10-C separately to train a Vision Transformer. The goal of this case study is to investigate and understand how the two models differ from each other through a loss landscape perspective. 

To begin, the analysts trained 3 checkpoints for each model with different random seeds. They see that the two models have different structures in the global structure view. By looking at the performance, the checkpoints of the altered ViT have higher test accuracy than those of the original ViT, and the altered ViT has better mode connectivity between checkpoints than the original ViT. This indicates that the loss landscape of the altered ViT seems to have a globally better-connected loss landscape compared to the original ViT. However, the similarity between checkpoints is lower than those of the original ViT, indicating that the altered ViT models tend to have larger variance than the original ViT models. Indeed, CIFAR10-C includes common image corruptions and perturbations, which could explain why the trained models perform differently.

Out of curiosity, the analysts are interested in what these two models' local structure looks like. According to the radial bar charts shown in the inner ring of the circles, they notice that, compared to the other checkpoints, the original ViT checkpoint 123 has the flattest local area (indicated by small Hessian eigenvalues), while the altered ViT checkpoint 123 has the sharpest local area (indicated by one larger Hessian eigenvalue). To further explore how they differ from each other on the local scale, the analysts can select the two checkpoints to show their local structure views. By looking at the loss contours, they observed one local minimum for each checkpoint. Although the persistence diagrams show subtle information, the merge tree for the original ViT checkpoint has more branches than the altered ViT checkpoint, which verified that the latter has a flatter local minimum, an indicator of better performance.

Finally, the analysts see that the transformer layers are different between the two trained models. From the prediction disparity view, the altered ViT performs better in general. Together, these observations, that the altered ViT has a flatter loss landscape, and that the checkpoints may be less similar, suggest that different types of corrupted data can somehow improve the model's generalizability. The original ViT, in contrast, performs worse, but this could perhaps be explained by the fact that the ViT was trained from scratch, and pre-trained models might display better performance.

# Architecture Alternations of Resnet-20

Our second case study illustrates how our framework can be used to investigate models with different architectures. The goal of this case study is to diagnose whether residual connections in ResNet-20 can cause significant changes in terms of the loss landscape. 

The analysts trained 4 checkpoints separately for Resnet-20 with and without residual connections. From the global structure view, the layouts appear similar in that both of them do not show an obvious pattern. This observation suggests that the models may not have an obvious global organization in high dimensional space.

First, the analysts notice that, although the average performance of the two models is similar (with accuracy around 90 to 92%), the mode connectivity of Resnet-20-R models is better than the mode connectivity of Resnet-20-NR models, indicated by thinner and more curved lines connecting the different checkpoints. The analysts wonder if removing the residual connections can significantly affect the loss landscape's smoothness, matching the intuition suggested by prior work. In terms of CKA similarity, the analysts noticed that the structure seems similar in that neither model has an obvious pattern or spatial organization. This similarity, encoded by the distance between circles in the two-dimensional graph representation, suggests that both models make similar predictions, which could be explained in part by the fact that both models were trained on the same data. The observations made based on the global structure views indicate that Resnet-20-R has a locally flat, globally well-connected landscape, and that some checkpoints are similar, while Resnet-20-NR has a locally flat, globally poorly-connected landscape.

It is noteworthy that Resnet-20-NR checkpoints 123 and 0 are very close to each other, yet their connectivity is relatively poor. This was not observed for Resnet-20-R, which suggests that the "high barrier" between Resnet-20-NR checkpoints is caused by the removal of the residual connections. 

Next, the analysts explore and compare the local structure for Resnet-20-R checkpoint 2023 and Resnet-20-NR checkpoint 2023. They notice that Resnet-20-R has "flat" loss landscapes compared to Resnet-20-NR. This is confirmed by the persistence diagram view, where Resnet-20-NR displays larger fluctuations compared to Resnet-20-R. The merge tree for Resnet-20-NR also shows many saddle points and local minima, whereas the merge tree for Resnet-20-R shows only one minimum. This suggests that the residual connections in the ResNet-20 architecture result in a better loss landscape both globally and locally. 

Finally, the analysts perform a semi-global analysis of the selected checkpoints. The layer similarity matrix indicates that the convolutional layers of Resnet-20-R tend to learn knowledge earlier than Resnet-20-NR, because the earlier layers in Resnet-20-NR are less similar than those of Resnet-20-R. This may be due to the residual connections which tend to yield better predictions, as shown in the prediction disparity view. This case study shows that even though the performance differences between models with and without residual connection is small, we can still see significant differences in the structure of their loss landscapes.

# Physical Constraints Incorporation of Models

In our third case study, we illustrate how our framework can be used to diagnose and analyze models where constraints are applied to the loss function during training. Specifically, we look at physics-informed neural networks (also know as PINNs). The goal of this study is to investigate the soft regularization used by PINNs by seeing how the loss landscape and prediction error changes as a physical parameter is varied. Here we look at PINNs trained to solve simple convection problems and we vary beta, a physical parameter corresponding to wave speed.

At first glance, the analysts notice that the checkpoints trained with beta=1 are closer to each other than those trained with beta=50. The shape formed by the set of checkpoints trained with beta=50 seems to be a "magnified" version of the shape formed by the set of checkpoints trained with beta=1. Accordingly, the mode connectivity between checkpoints trained with beta=50 is also poorer than those trained with beta=1, indicated by the thinner and more curved lines connecting the different checkpoints. Looking at their performance, the error rate is also higher for the higher value of beta. The analysts make a hypothesis that increasing beta leads to the model having a globally poorly connected and locally sharp loss landscape. 

To verify the hypothesis, the analysts select two checkpoints with different beta for further local structure inspection. To better investigate how the physical parameter influences the model, the analysts select checkpoint 123 for local landscape comparison. 

It turns out that the loss landscapes of the two models have many differences. The one trained with lower beta has a "flat" loss landscape and the other model trained with beta=50 does not have a very clear minimum. However, the persistence diagram shows that the terrain of beta=50 around the minimum is very rough and complex. It is also obvious that, from the merge tree, many more branches are observed for the model trained with beta=50 compared to the model trained with beta=1. Together, these insights suggest that increasing beta can lead to a poorly-connected, locally sharp loss landscape, thereby making the PINN model more prone to failure.

Finally, the analysts investigate model structure and prediction performance. The layer similarity view shows that the model starts to diverge in the third layer, and there are even more differences in the fourth layer. Such differences are also captured by the prediction disparity view, where the checkpoint trained with higher beta seems to have a larger prediction error. In conclusion, increasing beta causes the model to fail. Although we do not know exactly why this increase causes the model to fail, our experts expressed strong interest in conducting more experiments using LossLens for future research.
